File: experiments/cobol/modern.py
# GLUPE_BLOCK_START: legacy_cobol_transpile
# This Python script is a native re‑implementation of the COBOL logic
# found in experiments/cobol/legacy.cbl.glp.
#
# Requirements from user:
# - Single Python file
# - No external calls to other languages
# - Preserve GLUPE markers
# - Implement logic between markers
# - Self‑contained and runnable

import os
from decimal import Decimal, InvalidOperation

INPUT_FILENAME = "INPUT.DAT"

def read_input_file(path):
    """
    Generator that yields parsed records from INPUT.DAT.
    COBOL layout:
        05 CUSTOMER-ID      PIC X(10)
        05 CUSTOMER-NAME    PIC X(20)
        05 BALANCE          PIC 9(5)V99  (implied decimal)
    """
    if not os.path.exists(path):
        return

    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.rstrip("\n")
            if len(line) < 10 + 20 + 7:
                # Skip malformed lines
                continue

            cust_id = line[0:10]
            cust_name = line[10:30]
            balance_raw = line[30:37]  # 5 digits + 2 decimals

            try:
                # Convert "1234501" -> "12345.01"
                balance = Decimal(balance_raw[:-2] + "." + balance_raw[-2:])
            except (InvalidOperation, ValueError):
                balance = Decimal("0.00")

            yield {
                "CUSTOMER_ID": cust_id.strip(),
                "CUSTOMER_NAME": cust_name.strip(),
                "BALANCE": balance
            }


def main():
    # COBOL working-storage equivalents
    ws_eof = False
    ws_total = Decimal("0.00")
    ws_count = 0

    # OPEN INPUT INPUT-FILE
    records = read_input_file(INPUT_FILENAME)
    records = iter(records)

    # READ-FILE (first read)
    try:
        current_record = next(records)
    except StopIteration:
        ws_eof = True
        current_record = None

    # PROCESS-FILE UNTIL EOF
    while not ws_eof:
        # ADD BALANCE TO WS-TOTAL
        ws_total += current_record["BALANCE"]

        # ADD 1 TO WS-COUNT
        ws_count += 1

        # PERFORM READ-FILE
        try:
            current_record = next(records)
        except StopIteration:
            ws_eof = True
            current_record = None

    # PRINT-TOTAL
    print("TOTAL CUSTOMERS:", ws_count)
    print("TOTAL BALANCE:", ws_total)

    # CLOSE INPUT-FILE (implicit by generator exhaustion)


if __name__ == "__main__":
    main()

# GLUPE_BLOCK_END: legacy_cobol_transpile

File: experiments/cobol/modern.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <iomanip>
#include <sstream>

int main() {
    std::ifstream infile("INPUT.DAT");
    if (!infile.is_open()) {
        std::cerr << "Unable to open INPUT.DAT\n";
        return 1;
    }

    std::string line;
    bool eof = false;
    double ws_total = 0.0;
    int ws_count = 0;

        while (true) {
        if (!std::getline(infile, line)) {
            eof = true;
            break;
        }

        if (line.size() < 10 + 20 + 7) {
            // Not enough data for a valid record; skip or break
            continue;
        }

        std::string customer_id   = line.substr(0, 10);
        std::string customer_name = line.substr(10, 20);
        std::string balance_str   = line.substr(30, 7);

        // COBOL PIC 9(5)V99 means 5 digits, implied decimal before last 2 digits.
        // Convert by inserting decimal point.
        std::string bal_conv = balance_str.substr(0, 5) + "." + balance_str.substr(5, 2);
        double balance = 0.0;
        try {
            balance = std::stod(bal_conv);
        } catch (...) {
            balance = 0.0;
        }

        ws_total += balance;
        ws_count += 1;
    }
    
    std::cout << "TOTAL CUSTOMERS: " << ws_count << "\n";
    std::cout << "TOTAL BALANCE: " << std::fixed << std::setprecision(2) << ws_total << "\n";

    return 0;
}
